src/utils/tdex.ts

```diff
@@ -118,102 +54,130 @@ export async function calculatePrice(
+      });
     }
   }
 
   return trades;
 }
 
-/**
- * Filter a set of markets using asset to sent.
- * @param markets
- * @param sentAsset
- */
-export function getTradablesAssets(markets: TDEXMarket[], sentAsset: string): AssetWithTicker[] {
-  const results: AssetWithTicker[] = [];
+// Find all assets in markets tradable with the asset `asset`
+export function getTradablesAssets(markets: TDEXMarket[], asset: string): string[] {
+  const tradable: string[] = [];
 
   for (const market of markets) {
-    if (sentAsset === market.baseAsset && !results.map((r) => r.asset).includes(market.quoteAsset)) {
-      const mainAsset = getMainAsset(market.quoteAsset);
-      const ticker = mainAsset ? mainAsset.ticker : market.quoteAsset.slice(0, 4).toUpperCase();
-      const coinGeckoID = mainAsset?.coinGeckoID;
-
-      results.push({
-        asset: market.quoteAsset,
-        ticker,
-        coinGeckoID,
-      });
+    if (asset === market.baseAsset && !tradable.includes(market.quoteAsset)) {
+      tradable.push(market.quoteAsset);
     }
 
-    if (sentAsset === market.quoteAsset && !results.map((r) => r.asset).includes(market.baseAsset)) {
-      const mainAsset = getMainAsset(market.baseAsset);
-      const ticker = mainAsset ? mainAsset.ticker : market.baseAsset.slice(0, 4).toUpperCase();
-      const coinGeckoID = mainAsset?.coinGeckoID;
-
-      results.push({
-        asset: market.baseAsset,
-        ticker,
-        coinGeckoID,
-      });
+    if (asset === market.quoteAsset && !tradable.includes(market.baseAsset)) {
+      tradable.push(market.baseAsset);
     }
   }
 
-  return results;
+  return tradable;
 }
 
+export const assetHashToAssetConfig =
+  (assetRegistry: Record<string, AssetConfig>) =>
+  (assetHash: string): AssetConfig =>
+    assetRegistry[assetHash];
+
 const TDexRegistryURL = 'https://raw.githubusercontent.com/TDex-network/tdex-registry/master/registry.json';
 
 export async function getProvidersFromTDexRegistry(): Promise<TDEXProvider[]> {
   return (await axios.get(TDexRegistryURL)).data;
 }
+
+// discover the best order from a set of markets
+// return an async function
+export function discoverBestOrder(
+  markets: TDEXMarket[],
+  sendAsset?: string,
+  receiveAsset?: string
+): (sats: number, asset: string) => Promise<TradeOrder> {
+  if (!sendAsset || !receiveAsset) throw new Error('unable to compute orders for selected market');
+  const allPossibleOrders = computeOrders(markets, sendAsset, receiveAsset);
+  return async (sats: number, asset: string): Promise<TradeOrder> => {
+    if (sats <= 0) {
+      return allPossibleOrders[0];
+    }
+
+    try {
+      const discoverer = createDiscoverer(allPossibleOrders);
+      const bestOrders = await discoverer.discover({ asset, amount: sats });
+      if (bestOrders.length === 0) throw new Error('0 possible order for this trade');
+      return bestOrders[0];
+    } catch {
+      throw new Error('unable to find a valid TDEX order');
```

Comment:
not entirely clear what this message means for the end user

Reply:
instead of throwing an error, I finally return `allPossibleOrders[0]`: it means that if the discoverer fails to find the best order among all the possible orders, we just return the first one.